### PR TITLE
Lazier return in `UnionType->isSuperTypeOfWithReason()` (~40% faster)

### PR DESCRIPTION
**Pull request by staabm**
slow profile before this PR of https://github.com/phpstan/phpstan/issues/12159#issuecomment-2508567762 with https://github.com/phpstan/phpstan-src/commit/14d2bed182eb17346fd457f2c798301b17c42856

grafik

---

this PR:
```
➜  phpstan-src git:(lazy-or-1x) ✗ hyperfine 'php bin/phpstan analyze test.php --debug' -i
Benchmark 1: php bin/phpstan analyze test.php --debug
  Time (mean ± σ):      5.175 s ±  0.028 s    [User: 5.041 s, System: 0.131 s]
  Range (min … max):    5.137 s …  5.218 s    10 runs

```

with 1.12.x on https://github.com/phpstan/phpstan-src/commit/8d67d7a7617fcc3fb88dc5544f281d288ad696f4
```
➜  phpstan-src git:(1.12.x) ✗ hyperfine 'php bin/phpstan analyze test.php --debug' -i
Benchmark 1: php bin/phpstan analyze test.php --debug
  Time (mean ± σ):      8.579 s ±  0.080 s    [User: 8.424 s, System: 0.148 s]
  Range (min … max):    8.493 s …  8.727 s    10 runs

```

### CODE REVIEW

File: src/Type/UnionType.php [hunk 1]
```diff
@@ -245,7 +245,16 @@ public function isSuperTypeOfWithReason(Type $otherType): IsSuperTypeOfResult
...
 		}
 
-		$result = IsSuperTypeOfResult::createNo()->or(...array_map(static fn (Type $innerType) => $innerType->isSuperTypeOfWithReason($otherType), $this->types));
+		$results = [];
+		foreach ($this->types as $innerType) {
+			$result = $innerType->isSuperTypeOfWithReason($otherType);
+			if ($result->yes()) {
+				return $result;
+			}
+			$results[] = $result;
```
Comment (staabm) [inline src/Type/UnionType.php:254]: logic analog `TrinaryLogic->lazyOr()`

File: src/Type/UnionType.php [hunk 2]
```diff
@@ -245,7 +245,16 @@ public function isSuperTypeOfWithReason(Type $otherType): IsSuperTypeOfResult
...
+		}
+
+		$result = IsSuperTypeOfResult::createNo()->or(...$results);
 		if ($result->yes()) {
```
Comment (ondrejmirtes) [inline src/Type/UnionType.php]: It cannot be a yes after this, righr. It could be deleted.

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]

File: src/Type/UnionType.php (hunk 2 above)
Comment (staabm) [reply src/Type/UnionType.php]: right

**Comment by ondrejmirtes** [comment]
Now please send the merge commit.

**Pull request merged**